Variational autoencoder for tract profile data

SamChou05 opened the pull request.

SamChou05 committed:
fixing line lengths

SamChou05 committed:
Merge branch 'tractometry:main' into main

SamChou05 committed:
plotting latent space

SamChou05 committed:
change decoder

SamChou05 committed:
removing unecessary file

SamChou05 committed:
removing file

SamChou05 committed:
moving code to pt_models

SamChou05 committed:
temporary adding for debugging

SamChou05 committed:
making batchsize alterable

SamChou05 committed:
adding linear autoencoders

SamChou05 committed:
removing temp file

SamChou05 committed:
adding comments for future use

SamChou05 committed:
smoke tests for linear autoencoders

SamChou05 committed:
updating class names

SamChou05 committed:
condensing tests

SamChou05 committed:
adding conv autoencoder

SamChou05 committed:
one tract autoencoder

SamChou05 committed:
adding fit and fit transform

SamChou05 committed:
updated training loop

arokem commented:
@SamChou05 : now that #19 is merged, please go ahead and merge the main branch into your branch or rebase on top of main so that we can see whether your tests will pass.

SamChou05 committed:
test

SamChou05 committed:
adding less tract autoencode

SamChou05 committed:
Merge branch 'tractometry:main' into vae

SamChou05 committed:
starting to shift and modularize experiments ran on autoencodoers

SamChou05 committed:
adding linear one tract autoencoder for experiemnts

SamChou05 committed:
fixing ruffs

SamChou05 committed:
cleaning up code

SamChou05 committed:
fixing ruffs

SamChou05 committed:
comment too long

SamChou05 committed:
updated comment

SamChou05 committed:
made into a method

SamChou05 committed:
renaming for clarity

SamChou05 committed:
more renaming for clarity

SamChou05 committed:
updating renaming

SamChou05 committed:
updaing testa dn renaming

SamChou05 committed:
reducing pt models redundant code

SamChou05 committed:
reducing redundant code

SamChou05 committed:
trimming down test code

SamChou05 committed:
adding first tract fit to autoencoder

SamChou05 committed:
removing expriement code and moving to separate repo

SamChou05 committed:
thinning pt model code

SamChou05 committed:
cleaning up code

SamChou05 committed:
adding first tract memthod

SamChou05 committed:
adding back len to class

SamChou05 committed:
latent dim to latent dims

SamChou05 committed:
latent dim to latent dims

SamChou05 committed:
updating tests for autoencoders

SamChou05 committed:
changing device

SamChou05 committed:
changing device

SamChou05 committed:
changing autoencoder architecture

SamChou05 committed:
testing something

SamChou05 committed:
testing something

SamChou05 committed:
testing something

SamChou05 committed:
cleaning up code

SamChou05 committed:
fixing returns

SamChou05 committed:
fixing autoencoder FC

SamChou05 committed:
simplifying fits and fixing transforms for each autoencoder

SamChou05 committed:
cleaning up code

SamChou05 committed:
cleanign up code

SamChou05 committed:
fixing tests

arokem commented on afqinsight/nn/utils.py:
```diff
@@ -427,10 +429,107 @@ def prep_pytorch_data(dataset):
+    """
+    fa_indices = []
+    for i, fname in enumerate(dataset.feature_names):
+        if fname[0] == "dki_fa":
```
This label may vary between datasets

SamChou05 committed:
fixing padding issue

SamChou05 committed:
adding fa flattened method, renaming, and adding comments

SamChou05 committed:
updated variational auteoncoders

SamChou05 committed:
added utilities to accompany better variational autoencoders

SamChou05 committed:
changing convolutional variational archetecture

SamChou05 committed:
no need for recon loss anymore

SamChou05 committed:
separating decoder and variatonal conv decoder

SamChou05 committed:
fixing model arch

SamChou05 committed:
adding history for saving info

SamChou05 committed:
fixing otuput padding issue

SamChou05 committed:
adding notes

arokem commented on afqinsight/nn/pt_models.py:
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
 
 
+class VariationalEncoder(nn.Module):
+    def __init__(self, input_shape, latent_dims=20, dropout=0.2):
```
Looks like the `__init__` for this and the subsequent `Encoder` class have a lot in common. Are they actually identical, except for variable names? If so, maybe they can both rely on the same base class? Or one be the base and the other inherits from it?

arokem commented on afqinsight/nn/pt_models.py:
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
+        # mu = self.conv_mu(x)
+        # logvar = self.conv_logvar(x)
+
+        x = self.flatten(x)  # [64, 64*7]
```
What does this comment mean?

SamChou05 replied:
It was a note for the shape of the data after flattening in one of the experiments I was running, getting rid of it as it shouldn't be there.

arokem commented on afqinsight/nn/pt_models.py:
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
+        x = self.relu(self.conv3(x))
+        x = self.dropout(x)
+
+        # mu = self.conv_mu(x)
```
Remove?

arokem commented on afqinsight/nn/pt_models.py:
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
+        self.relu = nn.ReLU()
+
+    def forward(self, x):
+        # x = F.relu(self.deconv1(x))
```
Remove?

arokem commented on afqinsight/nn/pt_models.py:
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
+        return x
+
+
+class Conv1DVariationalEncoder(nn.Module):
```
Same comment as above. Could this be consolidated with the Conv1DEncoder?

arokem commented on afqinsight/nn/pt_models.py:
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
+        train_kl_per_epoch = []
+        train_recon_per_epoch = []
+
+        beta_start = 0.0
```
```suggestion
        # This is the weight for the KL "annealing"
        beta_start = 0.0
```

arokem commented on afqinsight/nn/pt_models.py:
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
+            running_recon_loss = 0.0
+            items = 0
+
+            beta = beta_start + slope * epoch
```
```suggestion
            # The weight on KL changes linearly with epoch:
            beta = beta_start + slope * epoch
```

arokem commented on afqinsight/nn/pt_models.py:
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
+            train_kl_per_epoch.append(avg_train_kl)
+            train_recon_per_epoch.append(avg_train_recon_loss)
+
+            print(
```
Maybe use `logging` instead of `print`?

SamChou05 committed:
removing unecessary comments

SamChou05 committed:
changing mu to mean for clarity

SamChou05 committed:
Update afqinsight/nn/pt_models.py comment for beta annealing

Co-authored-by: Ariel Rokem <arokem@gmail.com>

SamChou05 committed:
Update afqinsight/nn/pt_models.py updating comment for beta annealing

Co-authored-by: Ariel Rokem <arokem@gmail.com>

SamChou05 committed:
final string fix

SamChou05 committed:
Merge remote-tracking branch 'origin/vae' into vae
 the commit.

SamChou05 committed:
refactoring encoder classes

SamChou05 committed:
changing prep_fa_dataset for more customizability

arokem commented:
The code in pt_models looks great! 

I wonder whether `prep_first_tract_data`, `prep_fa_dataset`, and `prep_fa_flattned_data` are all a bit specific to the analysis that you are currently doing, so maybe they belong in your analysis code repo instead of here?

SamChou05 committed:
removing experiment - specific code for experiemnts

SamChou05 committed:
removing unecessary commits

SamChou05 commented:
I agree with your comment with prep_first_tract_data, prep_fa_dataset, and prep_fa_flattned_data, so I moved them to my repository containing the experiments and removed them from this PR.

arokem commented:
Great! You'll need to move out the related test as well

SamChou05 committed:
fixing tests

SamChou05 committed:
changed convolutional models to be more adaptable to more situations

SamChou05 committed:
adding changes to non variational conv

SamChou05 committed:
making code more consistent

arokem commented:
Could you please add docstrings to all the new functions/classes?

SamChou05 committed:
adding docstrings
